### PR TITLE
PMT Frontend webapp

### PR DESCRIPTION
**Pull request by JackUrb**
This PR lays the groundwork for the frontend webapp that will eventually be used to test UIs, launch tasks, manage qualifications, review work, manage workers, and more.

Right now it hooks up to the DB that is created when using the optional `use_db` flag. It can be used by running `npm install; npm run dev; python server.py` from inside the `webapp` directory.

Current features:
Provides views into runs and workers (currently it acts as a passive way to view metadata about runs).

Next up I'll be adding the reviewer interface, which will follow a change to how PMT stores data.

### CODE REVIEW

File: parlai/mturk/core/mturk_data_handler.py [hunk 1]
```diff
@@ -443,6 +470,16 @@ def get_worker_assignment_pairing(self, worker_id, assignment_id):
             results = c.fetchone()
             return results
 
+    def get_all_run_data(self, start=0, count=30):
+        '''get all the run data for all task_group_ids.'''
```
Comment (emilydinan) [inline parlai/mturk/core/mturk_data_handler.py]: nit: for all of our docstrings elsewhere we use double quotation marks

File: parlai/mturk/webapp/server.py [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+
+    def get(self):
+        try:
+            req = tornado.escape.json_decode(
+                tornado.escape.to_basestring(self.request.body)
+            )
+        except Exception:
+            req = {}
+
+        # self.render(
```
Comment (emilydinan) [inline parlai/mturk/webapp/server.py]: left in by accident?

File: parlai/mturk/webapp/server.py [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+                        help='host to run the server on.')
+    parser.add_argument('-db_file', metavar='db_file', type=str,
+                        default=DEFAULT_DB_FILE,
+                        help='name of database to use (in core/run_data)')
+    parser.add_argument('-logging_level', metavar='logger_level', default='INFO',
+                        help='logging level (default = INFO). Can take logging '
+                             'level name or int (example: 20)')
+    FLAGS = parser.parse_args()
+
+    try:
```
Comment (emilydinan) [inline parlai/mturk/webapp/server.py]: i'm pretty sure a call to _checkLevel already typechecks for you, so you can just always call that (and you don't neeed this try/except block

File: parlai/mturk/webapp/server.py [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+
+
+def get_rand_id():
+    return str(hex(int(time.time() * 10000000))[2:])
+
+
+def ensure_dir_exists(path):
+    """Make sure the parent dir exists for path so we can write a file."""
+    try:
+        os.makedirs(os.path.dirname(path))
```
Comment (klshuster) [inline parlai/mturk/webapp/server.py]: are you going for `os.makedirs(os.path.dirname(path), exists_ok=True)`?

File: parlai/mturk/webapp/server.py [hunk 5]
```diff
@@ -0,0 +1,375 @@
...
+        self.subs = app.subs
+        self.sources = app.sources
+        self.port = app.port
+        self.data_handler = app.data_handler
+
+    def get(self):
+        results = self.data_handler.get_all_worker_data()
+        processed_results = []
+        for res in results:
+            processed_results.append(dict(zip(res.keys(), res)))
```
Comment (klshuster) [inline parlai/mturk/webapp/server.py]: looks like you could use your `row_to_dict` function from above for `res`

File: parlai/mturk/core/mturk_data_handler.py (hunk 1 above)
Comment (JackUrb) [reply parlai/mturk/core/mturk_data_handler.py]: Oops, good catch. Fixing!

File: parlai/mturk/webapp/server.py (hunk 2 above)
Comment (JackUrb) [reply parlai/mturk/webapp/server.py]: Yup - fixed

File: parlai/mturk/core/mturk_data_handler.py [hunk 6]
```diff
@@ -473,6 +519,15 @@ def get_pairings_for_assignment(self, assignment_id):
             results = c.fetchall()
             return results
 
+    def get_pairings_for_run(self, task_group_id):
```
[the remark this answers is not in the document]
Comment (JackUrb) [reply parlai/mturk/core/mturk_data_handler.py]: Unintentional, adding docstrings

File: parlai/mturk/webapp/dev/app.jsx [hunk 7]
```diff
@@ -0,0 +1,805 @@
...
+      case 'failed':
+      default:
+        return <span>{props.value}</span>;
+    }
+  }
+
+  getColumnValue(header_name, item) {
+    switch(header_name) {
+      case 'run_id': return item.run_id;
+      case 'run_status': return 'complete';
```
[the remark this answers is not in the document]
Comment (JackUrb) [reply parlai/mturk/webapp/dev/app.jsx]: I should put a TODO here - I haven't been able to write the part that actually computes this value for a HIT yet

File: parlai/mturk/webapp/server.py (hunk 4 above)
Comment (JackUrb) [reply parlai/mturk/webapp/server.py]: Oh that works a lot better than this. Thanks for the note!

**Pull request merged**